### PR TITLE
Lazily define interface objects on globals (fixes #6419)

### LINKED ISSUE #6419: Stop defining all interface objects eagerly.
This will need a resolve hook, probably.

### PR DESCRIPTION
**Pull request by nox**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9652)

### CODE REVIEW

File: components/script/dom/browsingcontext.rs [hunk 1]
```diff
@@ -12,15 +12,15 @@ use dom::document::Document;
...
-use js::jsapi::{JSContext, JSErrNum, JSObject, JSPropertyDescriptor};
-use js::jsapi::{JS_AlreadyHasOwnPropertyById, JS_ForwardGetPropertyTo};
-use js::jsapi::{JS_DefinePropertyById6, JS_GetOwnPropertyDescriptorById};
-use js::jsval::{ObjectValue, UndefinedValue, PrivateValue};
+use js::glue::{CreateWrapperProxyHandler, GetProxyPrivate, NewWindowProxy};
+use js::glue::{ProxyTraps, SetProxyExtra};
+use js::jsapi::{Handle, HandleId, HandleObject, JS_DefinePropertyById6};
+use js::jsapi::{JS_ForwardGetPropertyTo, JS_ForwardSetPropertyTo, JS_GetClass};
+use js::jsapi::{JS_GetOwnPropertyDescriptorById, JS_HasPropertyById};
+use js::jsapi::{JSAutoCompartment, JSAutoRequest, JSContext, JSErrNum, JSObject};
```
Comment (KiChjang) [inline components/script/dom/browsingcontext.rs]: Tidy hates this line.

**Comment by nox** [comment]
These are intermittent.

**Comment by nox** [comment]
Blocked by #9728.

**Comment by nox** [comment]
https://github.com/sfackler/rust-phf/pull/80

**Comment by nox** [comment]
@Ms2ger This can be reviewed now.

**Comment by Ms2ger** [comment]
```js
window["abc\udc88"]
```
```
Assertion failure: hasLatin1Chars(), at /js/src/vm/String.h:1407
Servo exited with return value -11
```

**Comment by nox** [comment]
Guess I failed somewhere in the conversion code. Any other remark while at it?

**Comment by nox** [comment]
@Ms2ger I fixed this and added a test for it.

**Comment by Ms2ger** [comment]
Reviewed 6 of 12 files at r1, 2 of 2 files at r2, 9 of 11 files at r3.
Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 6049 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9652#-KBNXsZ_tKYxICB8rNGT:-KBNXsZ_tKYxICB8rNGU:1963125249) ([raw file](https://github.com/servo/servo/blob/5e4f5e666388209a5f609712f869a1c6faa92481/components/script/dom/bindings/codegen/CodegenRust.py#L6049)):*
Second time I see `d.interface.isCallback() or d.hasDescendants()`; worth creating a well-named method for it.

---

*[tests/wpt/mozilla/tests/mozilla/lazy-interface-object.html, line 15 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9652#-KBNS1TAIh2oJlgnrC8n:-KBNS1TAIh2oJlgnrC8o:-725124227) ([raw file](https://github.com/servo/servo/blob/5e4f5e666388209a5f609712f869a1c6faa92481/tests/wpt/mozilla/tests/mozilla/lazy-interface-object.html#L15)):*
I'd rather not rely on `TestBinding` and put this upstream.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9652)*

**Comment by Ms2ger** [comment]
Reviewed 13 of 13 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/bindings/utils.rs, line 391 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9652#-KBNpRgAmOzREp1-NMhN:-KBNpRgAmOzREp1-NMhO:-301787352) ([raw file](https://github.com/servo/servo/blob/95e375eb32cc9071b92ad5ed1743d27ebe1989c5/components/script/dom/bindings/utils.rs#L391)):*
Do this after the `JS_ResolveStandardClass` call, please.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9652)*

**Comment by nox** [comment]
@Ms2ger Done.

**Comment by Ms2ger** [comment]
@bors-servo r+

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9652#-:-KBNzym7iCJmVADb3aun:1236647940)*

**Pull request merged**